exec/cnex/lib/runtime_win32.c

```diff
@@ -0,0 +1,389 @@
+            return FALSE;
+        case DISP_E_UNKNOWNINTERFACE:
+            snprintf(err, sizeof(err), "unknown interface");
+            exec->rtl_raise(exec, "NativeObjectException", err);
+            return FALSE;
+        case DISP_E_UNKNOWNLCID:
+            snprintf(err, sizeof(err), "unknown lcid");
+            exec->rtl_raise(exec, "NativeObjectException", err);
+            return FALSE;
+        case DISP_E_PARAMNOTOPTIONAL:
+            snprintf(err, sizeof(err), "a required parameter was omitted");
+            exec->rtl_raise(exec, "NativeObjectException", err);
+            return FALSE;
+        default:
+            snprintf(err, sizeof(err), "unknown error from IDispatch::Invoke (0x%08X)", r);
+            exec->rtl_raise(exec, "NativeObjectException", err);
+            return FALSE;
+    }
+    return TRUE;
+}
+
+
+
+void com_releaseObject(Object *o)
+{
+    if (o != NULL) {
+        assert(o->refcount > 0);
+        o->refcount--;
+
+        if (o->refcount <= 0) {
+            if (o->ptr != NULL) {
+                free(((ComObject*)o->ptr)->name);
+                IDispatch_Release(((ComObject*)o->ptr)->obj);
+            }
+            free(o->ptr);
+            free(o);
+        }
+    }
+}
+
+Cell *com_toString(Object *self)
+{
+    char s[256];
+    snprintf(s, sizeof(s), "<ComObject:%s>", ((ComObject*)self->ptr)->name);
+    Cell *r = cell_fromCString(s);
+    return r;
+}
+
+BOOL com_invokeMethod(Object *self, TExecutor *exec, char *methodname, Cell *args, Cell **result)
+{
+    IDispatch *com = ((ComObject*)self->ptr)->obj;
+    BSTR mname = bstr_from_utf8string(methodname);
+    OLECHAR *a = mname;
+    DISPID dispid;
+    HRESULT r = com->lpVtbl->GetIDsOfNames(com, &IID_NULL, &a, 1, GetUserDefaultLCID(), &dispid);
+    SysFreeString(mname);
+    if (r != S_OK) {
+        char err[512];
+        snprintf(err, sizeof(err), "method not found in object (%s)", methodname);
+        exec->rtl_raise(exec, "NativeObjectException", err);
+        return FALSE;
+    }
+    DISPPARAMS params = { NULL, NULL, 0, 0 };
+    Array *pargs = args->array;
+    params.cArgs = (UINT)pargs->size;
+    params.rgvarg = malloc(sizeof(VARIANTARG) * params.cArgs);
+    for (size_t i = 0; i < pargs->size; i++) {
+        VARIANT *v = &params.rgvarg[params.cArgs - 1 - i];
+        VariantInit(v);
+        if (!VariantFromObject(exec, v, pargs->data[i].object)) {
+            return FALSE;
+        }
+    }
+    VARIANT rv;
+    VariantInit(&rv);
+    EXCEPINFO ei;
+    UINT argerr;
+    r = com->lpVtbl->Invoke(com, dispid, &IID_NULL, GetUserDefaultLCID(), DISPATCH_METHOD, &params, &rv, &ei, &argerr);
+    for (size_t i = 0; i < pargs->size; i++) {
+        VariantClear(&params.rgvarg[i]);
+    }
+    free(params.rgvarg);
+    if (!handleInvokeResult(exec, r, params.cArgs, &ei, argerr)) {
+        return FALSE;
+    }
+
+    ((Cell*)*result)->object = ObjectFromVariant(exec, &rv);
```

Comment:
There should not need to be a cast to `(Cell*)` here, because that's what the pointer already is. Remove the casts.

Reply:
Fixed.